pretyping/coercion.ml

```diff
@@ -367,36 +353,87 @@ let saturate_evd env evd =
   Typeclasses.resolve_typeclasses
     ~filter:Typeclasses.no_goals ~split:true ~fail:false env evd
 
+type coercion_trace =
+  | IdCoe
+  | Coe of {
+      isproj : Projection.Repr.t option;
+      head : econstr;
+      args : econstr list;
+      previous : coercion_trace;
+    }
+  | ProdCoe of { na : Name.t binder_annot; ty : econstr; dom : coercion_trace; body : coercion_trace }
```

Comment:
I think it would be more clear if `head` and `isproj` were merged in a single data type, either a term and some arguments or a projection and less arguments. It would avoid the skipn and the assert false.

Reply:
>It would avoid the skipn and the assert false.

Wouldn't they move to where we build the `Coe` from here where we destruct it?

Reply:
Yes, there is no magic, but changing the data as I suggest makes it harder to misuse it, since in order to access the arguments and the head you have to match and the data you find in the branch is already good. As it is now one can ignore isproj and still access head+args, but that would be wrong since args may need to be patched.

Reply:
Done.